testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java

```diff
@@ -0,0 +1,268 @@
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.TopicPartition;
+
+import com.google.protobuf.ByteString;
+
+import com.linecorp.decaton.processor.DecatonProcessor;
+import com.linecorp.decaton.processor.ProcessorProperties;
+import com.linecorp.decaton.processor.ProcessorsBuilder;
+import com.linecorp.decaton.processor.PropertySupplier;
+import com.linecorp.decaton.processor.runtime.ProcessorSubscription;
+import com.linecorp.decaton.processor.runtime.RetryConfig;
+import com.linecorp.decaton.protocol.Decaton.DecatonTaskRequest;
+import com.linecorp.decaton.protocol.Decaton.TaskMetadataProto;
+import com.linecorp.decaton.testing.KafkaClusterRule;
+import com.linecorp.decaton.testing.TestUtils;
+import com.linecorp.decaton.testing.processor.ProcessingGuarantee.GuaranteeType;
+import com.linecorp.decaton.testing.processor.TestTask.TestTaskDeserializer;
+import com.linecorp.decaton.testing.processor.TestTask.TestTaskSerializer;
+
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import lombok.experimental.Accessors;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Test suite that checks Decaton's core functionality, especially for processing semantics.
+ *
+ * Test suite will perform following procedure:
+ *   1. Start multiple subscription instances
+ *   2. Produce tasks
+ *   3. When half of tasks are processed, restart subscriptions one by one
+ *   4. Await all produced offsets are committed
+ *   5. Assert processing semantics
+ *
+ * Expected semantics can be customized according to each test instance.
+ * (e.g. process ordering will be no longer kept if the processor uses retry-queueing feature)
+ */
+@Slf4j
+@RequiredArgsConstructor
+public class ProcessorTestSuite {
+    private final KafkaClusterRule rule;
+    private final Function<ProcessorsBuilder<TestTask>, ProcessorsBuilder<TestTask>> configureProcessorsBuilder;
+    private final RetryConfig retryConfig;
+    private final PropertySupplier propertySuppliers;
+    private final EnumSet<GuaranteeType> semantics;
+
+    private static final int NUM_TASKS = 10000;
+    private static final int NUM_KEYS = 100;
+    private static final int NUM_SUBSCRIPTION_INSTANCES = 3;
+    private static final int NUM_PARTITIONS = 8;
+
+    @Setter
+    @Accessors(fluent = true)
+    public static class Builder {
+        private final KafkaClusterRule rule;
+
+        /**
+         * Configure test-specific processing logic
+         */
+        private Function<ProcessorsBuilder<TestTask>, ProcessorsBuilder<TestTask>> configureProcessorsBuilder;
+        /**
+         * Configure retry-queueing feature for the subscription
+         */
+        private RetryConfig retryConfig;
```

Comment:
This builder also allows to custom `ProcessorsBuilder` and `PropertySupplier`.

If we pass partially-built `SubscriptionBuilder`, users have to be carefully not to overwrite `processorsBuilder` because of following reasons. So it's not a good I/F I think?

- we're setting `preprocessor` as head of processor pipeline to record processing time
- supplied `processorsBuilder` has to consume `TestTask`

Reply:
Hm, ok sound good so far then.
Actually I was thinking that the current CoreFunctionalityTest should also run the same test with configuring processors with 2 different scopes "partition" and "thread' respectively, and I thought its required to access SubscriptionBuilder directly but just turns out that its actually configuration of `ProcessorsBuilder`.